src/Microsoft.Framework.FileSystemGlobbing/Infrastructure/MatcherContext.cs

```diff
@@ -3,275 +3,180 @@
 
 using System.Collections.Generic;
 using System.IO;
+using System.Linq;
 using Microsoft.Framework.FileSystemGlobbing.Abstractions;
 
 namespace Microsoft.Framework.FileSystemGlobbing.Infrastructure
 {
     public class MatcherContext
     {
+        private DirectoryInfoBase _root;
+        private IList<PatternContextBase> _includePatternContexts;
+        private IList<PatternContextBase> _excludePatternContexts;
+
         public MatcherContext(Matcher matcher, DirectoryInfoBase directoryInfo)
         {
-            Matcher = matcher;
-            DirectoryInfo = directoryInfo;
+            _root = directoryInfo;
+
+            _includePatternContexts = new List<PatternContextBase>();
             foreach (var pattern in matcher.IncludePatterns)
             {
-                if (pattern.Contains == null)
-                {
-                    IncludePatternContexts.Add(new PatternContextLinearInclude(this, pattern));
-                }
-                else
-                {
-                    IncludePatternContexts.Add(new PatternContextRaggedInclude(this, pattern));
-                }
+                _includePatternContexts.Add(pattern.Contains == null ?
+                    (PatternContextBase)new PatternContextLinearInclude(this, pattern) : new PatternContextRaggedInclude(this, pattern));
             }
+
+            _excludePatternContexts = new List<PatternContextBase>();
             foreach (var pattern in matcher.ExcludePatterns)
             {
-                if (pattern.Contains == null)
-                {
-                    ExcludePatternContexts.Add(new PatternContextLinearExclude(this, pattern));
-                }
-                else
-                {
-                    ExcludePatternContexts.Add(new PatternContextRaggedExclude(this, pattern));
-                }
+                _excludePatternContexts.Add(pattern.Contains == null ?
+                    (PatternContextBase)new PatternContextLinearExclude(this, pattern) : new PatternContextRaggedExclude(this, pattern));
             }
         }
 
-        public Matcher Matcher { get; }
-        public DirectoryInfoBase DirectoryInfo { get; }
-        public IList<PatternContextBase> IncludePatternContexts { get; } = new List<PatternContextBase>();
-        public IList<PatternContextBase> ExcludePatternContexts { get; } = new List<PatternContextBase>();
-        public List<string> Files { get; private set; }
-
-        public FrameData Frame;
-        public Stack<FrameData> FrameStack = new Stack<FrameData>();
-
         public PatternMatchingResult Execute()
         {
-            Files = new List<string>();
+            var files = new List<string>();
 
-            Frame.Stage = Stage.Complete;
-            PushFrame(Stage.Predicting, DirectoryInfo);
+            // Walk the file system in Non-recursive, pre order, tree traversal
 
-            while (Frame.Stage != Stage.Complete)
+            var stack = new Stack<FrameData>();
```

Comment:
The FrameStack as member, Frame as field, and FrameData as value type were to avoid allocations for working structures. Should switch back to that technique.